src/python/pants/core/util_rules/archive.py

```diff
@@ -19,33 +20,58 @@
 from pants.util.logging import LogLevel
 
 
-@dataclass(frozen=True)
-class MaybeExtractable:
-    """A newtype to work around rule graph resolution issues.
+# Note that updating this will impact the `archive` target defined in `core/target_types.py`.
+class ArchiveFormat(Enum):
+    TAR = "tar"
+    TGZ = "tgz"
+    TBZ2 = "tbz2"
+    TXZ = "txz"
+    ZIP = "zip"
 
-    Possibly https://github.com/pantsbuild/pants/issues/9320. Either way, we should fix the
-    underlying issue and get rid of this type.
-    """
 
-    digest: Digest
+# -----------------------------------------------------------------------------------------------
+# Find binaries to create/extract archives
+# -----------------------------------------------------------------------------------------------
 
+# TODO: Should this be configurable?
+SEARCH_PATHS = ("/usr/bin", "/bin", "/usr/local/bin")
 
-@dataclass(frozen=True)
-class ExtractedDigest:
-    """The result of extracting an archive."""
 
-    digest: Digest
+class ZipBinary(BinaryPath):
+    def create_archive_argv(self, request: "CreateArchive") -> Tuple[str, ...]:
+        return (self.path, request.output_filename, *request.snapshot.files)
 
 
-# TODO: Should this be configurable?
-SEARCH_PATHS = ("/usr/bin", "/bin", "/usr/local/bin")
+class UnzipBinary(BinaryPath):
+    def extract_archive_argv(self, *, archive_path: str, output_dir: str) -> Tuple[str, ...]:
+        # Note that the `output_dir` does not need to already exist.
+        # The caller should validate that it's a valid `.zip` file.
+        return (self.path, archive_path, "-d", output_dir)
 
 
-class UnzipBinary(BinaryPath):
-    def extract_argv(self, *, archive_path: str, output_dir: str) -> Tuple[str, ...]:
-        # Note that the `output_dir` does not need to already exist. The caller should also
-        # validate that it's a valid `.zip` file.
-        return (self.path, "-q", archive_path, "-d", output_dir)
+class TarBinary(BinaryPath):
+    def create_archive_argv(self, request: "CreateArchive") -> Tuple[str, ...]:
+        # Note that the parent directory for the output_filename must already exist.
+        #
+        # The `a` arg means `tar` will choose the compression based on the file ending.
```

Comment:
How widely available is the `a` option to tar? I see it in GNU tar and my latest macOS version BSD tar but it may not be available in every tar implementation.

Reply:
Good question.

>  For maximum portability, scripts that invoke tar should use the bundled-argument for-
     mat above, should limit themselves to the c, t, and x modes, and the b, f, m, v, and
     w options.

None of those options allow for creating .tgz, .txz, and .tbz2, though. We'd need to use the dedicated flags. I can't find if the dedicated flags are more compatible than the auto-flag.